lib/map.h

```diff
@@ -17,42 +17,66 @@ limitations under the License.
 #ifndef LIB_MAP_H_
 #define LIB_MAP_H_
 
-#include <map>
+#include <optional>
 
-template <class K, class T, class V, class Comp, class Alloc>
-inline V get(const std::map<K, V, Comp, Alloc> &m, T key, V def = V()) {
+/// Given a map and a key, return the value corresponding to the key in the map,
+/// or a given default value if the key doesn't exist in the map.
+template <typename Map, typename Key>
+typename Map::mapped_type get(const Map &m, const Key &key) {
     auto it = m.find(key);
-    if (it != m.end()) return it->second;
-    return def;
+    return it != m.end() ? it->second : typename Map::mapped_type{};
+}
+template <class Map, typename Key = typename Map::key_type,
+          typename Value = typename Map::mapped_type>
+typename Map::mapped_type get(const Map &m, const Key &key, Value &&def) {
+    using M = typename Map::mapped_type;
+    auto it = m.find(key);
+    return it != m.end() ? it->second : static_cast<M>(static_cast<Value &&>(def));
+}
+
+/// Given a map and a key, return the pointer to value corresponding to the key
+/// in the map, or a nullptr if the key doesn't exist in the map.
+template <typename Map, typename Key>
+const typename Map::mapped_type *getref(const Map &m, const Key &key) {
+    auto it = m.find(key);
+    return it != m.end() ? &it->second : nullptr;
 }
 
-template <class K, class T, class V, class Comp, class Alloc>
-inline V *getref(std::map<K, V, Comp, Alloc> &m, T key) {
+template <typename Map, typename Key>
+typename Map::mapped_type *getref(Map &m, const Key &key) {
     auto it = m.find(key);
-    if (it != m.end()) return &it->second;
-    return 0;
+    return it != m.end() ? &it->second : nullptr;
 }
 
-template <class K, class T, class V, class Comp, class Alloc>
-inline const V *getref(const std::map<K, V, Comp, Alloc> &m, T key) {
+// Given a map and a key, return a optional<V> if the key exists and None if the
+// key does not exist in the map.
+template <class Map, typename Key = typename Map::key_type>
+std::optional<typename Map::mapped_type> get_optional(const Map &m, const Key &key) {
```

Comment:
`getopt` would be more consistent with `getref`. I'm not sure myself which one is better, we could go with the more explicit name you have chosen.

Reply:
I though about this. But `getopt` in a global scope is already taken by...getopt :) https://man7.org/linux/man-pages/man3/getopt.3.html So, less name clashes. It has different signature, but since it's a C function I'm afraid of macroses. We can give `getopt` a try and see how it will go.

Reply:
I see. Now I'm slightly leaning towards `get_optional` to avoid possible confusion. The error messages could get weird with `getopt` too.